# PROCESSING DATA WITH PANDAS

In [1]:
#import
import pandas as pd

In [2]:
#define file path
fp = "Kumpula-June-2016-w-metadata.txt"

#Read in the data file(starting ow 9)
data = pd.read_csv(fp, skiprows=8)

In [3]:
data

YEARMODA  TEMP   MAX   MIN
0   20160601  65.5  73.6  54.7
1   20160602  65.8  80.8  55.0
2   20160603  68.4   NaN  55.6
3   20160604  57.5  70.9  47.3
4   20160605  51.4  58.3  43.2
5   20160606  52.2  59.7  42.8
6   20160607  56.9  65.1  45.9
7   20160608  54.2   NaN  47.5
8   20160609  49.4  54.1  45.7
9   20160610  49.5  55.9  43.0
10  20160611  54.0  62.1  41.7
11  20160612  55.4  64.2  46.0
12  20160613  58.3  68.2  47.3
13  20160614  59.7  67.8  47.8
14  20160615  63.4  70.3  49.3
15  20160616  57.8  67.5  55.6
16  20160617  60.4  70.7  55.9
17  20160618  57.3   NaN  54.0
18  20160619  56.3  59.2  54.1
19  20160620  59.3  69.1  52.2
20  20160621  62.6  71.4  50.4
21  20160622  61.7  70.2  55.4
22  20160623  60.9  67.1  54.9
23  20160624  61.1  68.9  56.7
24  20160625  65.7  75.4  57.9
25  20160626  69.6  77.7  60.3
26  20160627  60.7  70.0   NaN
27  20160628  65.4  73.0  55.8
28  20160629  65.8  73.2   NaN
29  20160630  65.7  72.7  59.2

In [4]:
#confirm
data.head()

YEARMODA  TEMP   MAX   MIN
0  20160601  65.5  73.6  54.7
1  20160602  65.8  80.8  55.0
2  20160603  68.4   NaN  55.6
3  20160604  57.5  70.9  47.3
4  20160605  51.4  58.3  43.2

In [5]:
#define new column "DIFF"
data["DIFF"] = 0.0
data.head()

YEARMODA  TEMP   MAX   MIN  DIFF
0  20160601  65.5  73.6  54.7   0.0
1  20160602  65.8  80.8  55.0   0.0
2  20160603  68.4   NaN  55.6   0.0
3  20160604  57.5  70.9  47.3   0.0
4  20160605  51.4  58.3  43.2   0.0

In [6]:
data["DIFF"].dtype

dtype('float64')

In [7]:
#calculat max min difference
data["DIFF"] = data["MAX"] - data["MIN"]

#result
data.head()

YEARMODA  TEMP   MAX   MIN  DIFF
0  20160601  65.5  73.6  54.7  18.9
1  20160602  65.8  80.8  55.0  25.8
2  20160603  68.4   NaN  55.6   NaN
3  20160604  57.5  70.9  47.3  23.6
4  20160605  51.4  58.3  43.2  15.1

In [8]:
#create a column quickly
data["DIFF2"] = data["MAX"] - data["MIN"] 
data.head()

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2
0  20160601  65.5  73.6  54.7  18.9   18.9
1  20160602  65.8  80.8  55.0  25.8   25.8
2  20160603  68.4   NaN  55.6   NaN    NaN
3  20160604  57.5  70.9  47.3  23.6   23.6
4  20160605  51.4  58.3  43.2  15.1   15.1

In [9]:
#crete column converting temp faahr to celsius
data["TEMP_CELSIUS"] = (data["TEMP"] - 32.0) / (9 / 5) 

data.head()

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS
0  20160601  65.5  73.6  54.7  18.9   18.9     18.611111
1  20160602  65.8  80.8  55.0  25.8   25.8     18.777778
2  20160603  68.4   NaN  55.6   NaN    NaN     20.222222
3  20160604  57.5  70.9  47.3  23.6   23.6     14.166667
4  20160605  51.4  58.3  43.2  15.1   15.1     10.777778

In [10]:
#create column converting celius to kelvin
data["TEMP_KELVIN"] = data["TEMP_CELSIUS"] + 273
data.head()

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
0  20160601  65.5  73.6  54.7  18.9   18.9     18.611111   291.611111
1  20160602  65.8  80.8  55.0  25.8   25.8     18.777778   291.777778
2  20160603  68.4   NaN  55.6   NaN    NaN     20.222222   293.222222
3  20160604  57.5  70.9  47.3  23.6   23.6     14.166667   287.166667
4  20160605  51.4  58.3  43.2  15.1   15.1     10.777778   283.777778

In [11]:
#selecting first five rows using row index values
data[0:7]

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
0  20160601  65.5  73.6  54.7  18.9   18.9     18.611111   291.611111
1  20160602  65.8  80.8  55.0  25.8   25.8     18.777778   291.777778
2  20160603  68.4   NaN  55.6   NaN    NaN     20.222222   293.222222
3  20160604  57.5  70.9  47.3  23.6   23.6     14.166667   287.166667
4  20160605  51.4  58.3  43.2  15.1   15.1     10.777778   283.777778
5  20160606  52.2  59.7  42.8  16.9   16.9     11.222222   284.222222
6  20160607  56.9  65.1  45.9  19.2   19.2     13.833333   286.833333

In [12]:
#selecting temp column values on rows 0-5
selection = data.loc[0:5, "TEMP"] #this selects start and end values unlike lists
selection

0    65.5
1    65.8
2    68.4
3    57.5
4    51.4
5    52.2
Name: TEMP, dtype: float64

In [13]:
#selecting columns temp and temp_celsius on rows 0-5
selection = data.loc[0:5, ["TEMP", "TEMP_CELSIUS"]]
selection

TEMP  TEMP_CELSIUS
0  65.5     18.611111
1  65.8     18.777778
2  68.4     20.222222
3  57.5     14.166667
4  51.4     10.777778
5  52.2     11.222222

In [14]:
#get mean values(celsius) from the last seven days of june
selection = data.loc[23:29, "TEMP_CELSIUS"].mean()
selection

18.253968253968257

In [15]:
#selecting one row
row = data.loc[4]

In [16]:
#print one attribute from selected row
row["TEMP"]

51.4

In [17]:
#selecting single value based on row and column
data.at[4, "TEMP"]

51.4

.loc and .at are based on the axis labels - the names of columns and rows. Axis labels can be also something else than “traditional” index values. For example, datetime is commonly used as the row index. .iloc is another indexing operator which is based on integer value indices. Using .iloc, it is possible to refer also to the columns based on their index value. For example, data.iloc[0,0] would return 20160601 in our example data frame.

In [18]:
data.iloc[0:5:, 0:2]

YEARMODA  TEMP
0  20160601  65.5
1  20160602  65.8
2  20160603  68.4
3  20160604  57.5
4  20160605  51.4

In [19]:
data.iloc[0, 1]

65.5

In [20]:
data.iloc[-1]

YEARMODA        2.016063e+07
TEMP            6.570000e+01
MAX             7.270000e+01
MIN             5.920000e+01
DIFF            1.350000e+01
DIFF2           1.350000e+01
TEMP_CELSIUS    1.872222e+01
TEMP_KELVIN     2.917222e+02
Name: 29, dtype: float64

FILTERING AND UPDATING DATA

In [21]:
#check condition temp > 15
data["TEMP_CELSIUS"] > 15.0


0      True
1      True
2      True
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13     True
14     True
15    False
16     True
17    False
18    False
19     True
20     True
21     True
22     True
23     True
24     True
25     True
26     True
27     True
28     True
29     True
Name: TEMP_CELSIUS, dtype: bool

In [22]:
#select rows with temp > 15 
warm_temps = data.loc[data["TEMP_CELSIUS"] > 15]
warm_temps

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
0   20160601  65.5  73.6  54.7  18.9   18.9     18.611111   291.611111
1   20160602  65.8  80.8  55.0  25.8   25.8     18.777778   291.777778
2   20160603  68.4   NaN  55.6   NaN    NaN     20.222222   293.222222
13  20160614  59.7  67.8  47.8  20.0   20.0     15.388889   288.388889
14  20160615  63.4  70.3  49.3  21.0   21.0     17.444444   290.444444
16  20160617  60.4  70.7  55.9  14.8   14.8     15.777778   288.777778
19  20160620  59.3  69.1  52.2  16.9   16.9     15.166667   288.166667
20  20160621  62.6  71.4  50.4  21.0   21.0     17.000000   290.000000
21  20160622  61.7  70.2  55.4  14.8   14.8     16.500000   289.500000
22  20160623  60.9  67.1  54.9  12.2   12.2     16.055556   289.055556
23  20160624  61.1  68.9  56.7  12.2   12.2     16.166667   289.166667
24  20160625  65.7  75.4  57.9  17.5   17.5     18.722222   291.722222
25  20160626  69.6  77.7  60.3  17.4   17.4     20.888889   293.888889
26  20160627  60.7  70.0   NaN   NaN    NaN     15.944444   288.944444
27  20160628  65.4  73.0  55.8  17.2   17.2     18.555556   291.555556
28  20160629  65.8  73.2   NaN   NaN    NaN     18.777778   291.777778
29  20160630  65.7  72.7  59.2  13.5   13.5     18.722222   291.722222

In [23]:
#select rows > 15 from late june
warm_temps = data.loc[(data["TEMP_CELSIUS"] > 15) & (data["YEARMODA"] >= 20160615)] #& is used because we want two columns for true and false values
warm_temps

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
14  20160615  63.4  70.3  49.3  21.0   21.0     17.444444   290.444444
16  20160617  60.4  70.7  55.9  14.8   14.8     15.777778   288.777778
19  20160620  59.3  69.1  52.2  16.9   16.9     15.166667   288.166667
20  20160621  62.6  71.4  50.4  21.0   21.0     17.000000   290.000000
21  20160622  61.7  70.2  55.4  14.8   14.8     16.500000   289.500000
22  20160623  60.9  67.1  54.9  12.2   12.2     16.055556   289.055556
23  20160624  61.1  68.9  56.7  12.2   12.2     16.166667   289.166667
24  20160625  65.7  75.4  57.9  17.5   17.5     18.722222   291.722222
25  20160626  69.6  77.7  60.3  17.4   17.4     20.888889   293.888889
26  20160627  60.7  70.0   NaN   NaN    NaN     15.944444   288.944444
27  20160628  65.4  73.0  55.8  17.2   17.2     18.555556   291.555556
28  20160629  65.8  73.2   NaN   NaN    NaN     18.777778   291.777778
29  20160630  65.7  72.7  59.2  13.5   13.5     18.722222   291.722222

In [24]:
#reset index
warm_temps = warm_temps.reset_index(drop=True) #the (drop=True) so as to drop the old index from the data frame
warm_temps

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
0   20160615  63.4  70.3  49.3  21.0   21.0     17.444444   290.444444
1   20160617  60.4  70.7  55.9  14.8   14.8     15.777778   288.777778
2   20160620  59.3  69.1  52.2  16.9   16.9     15.166667   288.166667
3   20160621  62.6  71.4  50.4  21.0   21.0     17.000000   290.000000
4   20160622  61.7  70.2  55.4  14.8   14.8     16.500000   289.500000
5   20160623  60.9  67.1  54.9  12.2   12.2     16.055556   289.055556
6   20160624  61.1  68.9  56.7  12.2   12.2     16.166667   289.166667
7   20160625  65.7  75.4  57.9  17.5   17.5     18.722222   291.722222
8   20160626  69.6  77.7  60.3  17.4   17.4     20.888889   293.888889
9   20160627  60.7  70.0   NaN   NaN    NaN     15.944444   288.944444
10  20160628  65.4  73.0  55.8  17.2   17.2     18.555556   291.555556
11  20160629  65.8  73.2   NaN   NaN    NaN     18.777778   291.777778
12  20160630  65.7  72.7  59.2  13.5   13.5     18.722222   291.722222

In [25]:
#check mean temp using YEARMODA column
data["TEMP_CELSIUS"].loc[data["YEARMODA"] >= 20160624].mean()

18.253968253968257

In [26]:
warm_temps

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
0   20160615  63.4  70.3  49.3  21.0   21.0     17.444444   290.444444
1   20160617  60.4  70.7  55.9  14.8   14.8     15.777778   288.777778
2   20160620  59.3  69.1  52.2  16.9   16.9     15.166667   288.166667
3   20160621  62.6  71.4  50.4  21.0   21.0     17.000000   290.000000
4   20160622  61.7  70.2  55.4  14.8   14.8     16.500000   289.500000
5   20160623  60.9  67.1  54.9  12.2   12.2     16.055556   289.055556
6   20160624  61.1  68.9  56.7  12.2   12.2     16.166667   289.166667
7   20160625  65.7  75.4  57.9  17.5   17.5     18.722222   291.722222
8   20160626  69.6  77.7  60.3  17.4   17.4     20.888889   293.888889
9   20160627  60.7  70.0   NaN   NaN    NaN     15.944444   288.944444
10  20160628  65.4  73.0  55.8  17.2   17.2     18.555556   291.555556
11  20160629  65.8  73.2   NaN   NaN    NaN     18.777778   291.777778
12  20160630  65.7  72.7  59.2  13.5   13.5     18.722222   291.722222

# DEALING WITH MISSING DATA

In [27]:
#DROP no data values on min colmn
warm_temps_clean = warm_temps.dropna(subset=["MIN"]) 
#if you dont want to include sepcific columns, dont use the subset and you can also add multiple
warm_temps_clean 
#index 9 & 11 removed

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
0   20160615  63.4  70.3  49.3  21.0   21.0     17.444444   290.444444
1   20160617  60.4  70.7  55.9  14.8   14.8     15.777778   288.777778
2   20160620  59.3  69.1  52.2  16.9   16.9     15.166667   288.166667
3   20160621  62.6  71.4  50.4  21.0   21.0     17.000000   290.000000
4   20160622  61.7  70.2  55.4  14.8   14.8     16.500000   289.500000
5   20160623  60.9  67.1  54.9  12.2   12.2     16.055556   289.055556
6   20160624  61.1  68.9  56.7  12.2   12.2     16.166667   289.166667
7   20160625  65.7  75.4  57.9  17.5   17.5     18.722222   291.722222
8   20160626  69.6  77.7  60.3  17.4   17.4     20.888889   293.888889
10  20160628  65.4  73.0  55.8  17.2   17.2     18.555556   291.555556
12  20160630  65.7  72.7  59.2  13.5   13.5     18.722222   291.722222

Note that we replaced the original warm_temps variable with version where no data are removed. The .dropna() function, among other pandas functions can also be applied “inplace” which means that the function updates the DataFrame object and returns None:

warm_temps.dropna(subset=['MIN'], inplace=True)

In [28]:
#fill na values
warm_temps.fillna(-9999)
#be careful when doing this, might give you wrong info when analysing data

YEARMODA  TEMP   MAX     MIN    DIFF   DIFF2  TEMP_CELSIUS  TEMP_KELVIN
0   20160615  63.4  70.3    49.3    21.0    21.0     17.444444   290.444444
1   20160617  60.4  70.7    55.9    14.8    14.8     15.777778   288.777778
2   20160620  59.3  69.1    52.2    16.9    16.9     15.166667   288.166667
3   20160621  62.6  71.4    50.4    21.0    21.0     17.000000   290.000000
4   20160622  61.7  70.2    55.4    14.8    14.8     16.500000   289.500000
5   20160623  60.9  67.1    54.9    12.2    12.2     16.055556   289.055556
6   20160624  61.1  68.9    56.7    12.2    12.2     16.166667   289.166667
7   20160625  65.7  75.4    57.9    17.5    17.5     18.722222   291.722222
8   20160626  69.6  77.7    60.3    17.4    17.4     20.888889   293.888889
9   20160627  60.7  70.0 -9999.0 -9999.0 -9999.0     15.944444   288.944444
10  20160628  65.4  73.0    55.8    17.2    17.2     18.555556   291.555556
11  20160629  65.8  73.2 -9999.0 -9999.0 -9999.0     18.777778   291.777778
12  20160630  65.7  72.7    59.2    13.5    13.5     18.722222   291.722222

DATA TYPE CONVERSIONS

In [29]:
data["YEARMODA"].astype(float)

0     20160601.0
1     20160602.0
2     20160603.0
3     20160604.0
4     20160605.0
5     20160606.0
6     20160607.0
7     20160608.0
8     20160609.0
9     20160610.0
10    20160611.0
11    20160612.0
12    20160613.0
13    20160614.0
14    20160615.0
15    20160616.0
16    20160617.0
17    20160618.0
18    20160619.0
19    20160620.0
20    20160621.0
21    20160622.0
22    20160623.0
23    20160624.0
24    20160625.0
25    20160626.0
26    20160627.0
27    20160628.0
28    20160629.0
29    20160630.0
Name: YEARMODA, dtype: float64

In [30]:
data["YEARMODA"].astype(str)

0     20160601
1     20160602
2     20160603
3     20160604
4     20160605
5     20160606
6     20160607
7     20160608
8     20160609
9     20160610
10    20160611
11    20160612
12    20160613
13    20160614
14    20160615
15    20160616
16    20160617
17    20160618
18    20160619
19    20160620
20    20160621
21    20160622
22    20160623
23    20160624
24    20160625
25    20160626
26    20160627
27    20160628
28    20160629
29    20160630
Name: YEARMODA, dtype: object

Be careful with type conversions from floating point values to integers. The conversion simply drops the stuff to the right of the decimal point, so all values are rounded down to the nearest whole number. For example, 99.99 will be truncated to 99 as an integer, when it should be rounded up to 100.

Chaining the round and type conversion functions solves this issue as the .round(0).astype(int) command first rounds the values with zero decimals and then converts those values into integers.

In [31]:
print("Original values:")
data["TEMP"].head()

Original values:


0    65.5
1    65.8
2    68.4
3    57.5
4    51.4
Name: TEMP, dtype: float64

In [32]:
print("Truncated integer values:")
data["TEMP"].astype(int).head()

Truncated integer values:


0    65
1    65
2    68
3    57
4    51
Name: TEMP, dtype: int32

In [33]:
#to round up
data["TEMP"].head().round(0).astype(int)

0    66
1    66
2    68
3    58
4    51
Name: TEMP, dtype: int32

In [34]:
# Get unique celsius values
unique = data["TEMP"].unique()
unique

array([65.5, 65.8, 68.4, 57.5, 51.4, 52.2, 56.9, 54.2, 49.4, 49.5, 54. ,
       55.4, 58.3, 59.7, 63.4, 57.8, 60.4, 57.3, 56.3, 59.3, 62.6, 61.7,
       60.9, 61.1, 65.7, 69.6, 60.7, 65.4])

In [35]:
# unique values as list
list(unique)

[65.5,
 65.8,
 68.4,
 57.5,
 51.4,
 52.2,
 56.9,
 54.2,
 49.4,
 49.5,
 54.0,
 55.4,
 58.3,
 59.7,
 63.4,
 57.8,
 60.4,
 57.3,
 56.3,
 59.3,
 62.6,
 61.7,
 60.9,
 61.1,
 65.7,
 69.6,
 60.7,
 65.4]

In [36]:
# Number of unique values
unique_temps = len(unique)
print(f"There were {unique_temps} days with unique mean temperatures in June 2016.")

There were 28 days with unique mean temperatures in June 2016.


SORTING DATA

In [37]:
#sort dataframe ascending
data.sort_values(by="TEMP")

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
8   20160609  49.4  54.1  45.7   8.4    8.4      9.666667   282.666667
9   20160610  49.5  55.9  43.0  12.9   12.9      9.722222   282.722222
4   20160605  51.4  58.3  43.2  15.1   15.1     10.777778   283.777778
5   20160606  52.2  59.7  42.8  16.9   16.9     11.222222   284.222222
10  20160611  54.0  62.1  41.7  20.4   20.4     12.222222   285.222222
7   20160608  54.2   NaN  47.5   NaN    NaN     12.333333   285.333333
11  20160612  55.4  64.2  46.0  18.2   18.2     13.000000   286.000000
18  20160619  56.3  59.2  54.1   5.1    5.1     13.500000   286.500000
6   20160607  56.9  65.1  45.9  19.2   19.2     13.833333   286.833333
17  20160618  57.3   NaN  54.0   NaN    NaN     14.055556   287.055556
3   20160604  57.5  70.9  47.3  23.6   23.6     14.166667   287.166667
15  20160616  57.8  67.5  55.6  11.9   11.9     14.333333   287.333333
12  20160613  58.3  68.2  47.3  20.9   20.9     14.611111   287.611111
19  20160620  59.3  69.1  52.2  16.9   16.9     15.166667   288.166667
13  20160614  59.7  67.8  47.8  20.0   20.0     15.388889   288.388889
16  20160617  60.4  70.7  55.9  14.8   14.8     15.777778   288.777778
26  20160627  60.7  70.0   NaN   NaN    NaN     15.944444   288.944444
22  20160623  60.9  67.1  54.9  12.2   12.2     16.055556   289.055556
23  20160624  61.1  68.9  56.7  12.2   12.2     16.166667   289.166667
21  20160622  61.7  70.2  55.4  14.8   14.8     16.500000   289.500000
20  20160621  62.6  71.4  50.4  21.0   21.0     17.000000   290.000000
14  20160615  63.4  70.3  49.3  21.0   21.0     17.444444   290.444444
27  20160628  65.4  73.0  55.8  17.2   17.2     18.555556   291.555556
0   20160601  65.5  73.6  54.7  18.9   18.9     18.611111   291.611111
24  20160625  65.7  75.4  57.9  17.5   17.5     18.722222   291.722222
29  20160630  65.7  72.7  59.2  13.5   13.5     18.722222   291.722222
1   20160602  65.8  80.8  55.0  25.8   25.8     18.777778   291.777778
28  20160629  65.8  73.2   NaN   NaN    NaN     18.777778   291.777778
2   20160603  68.4   NaN  55.6   NaN    NaN     20.222222   293.222222
25  20160626  69.6  77.7  60.3  17.4   17.4     20.888889   293.888889

In [38]:
#sort dataframe descending order
data.sort_values(by="TEMP", ascending=False)

YEARMODA  TEMP   MAX   MIN  DIFF  DIFF2  TEMP_CELSIUS  TEMP_KELVIN
25  20160626  69.6  77.7  60.3  17.4   17.4     20.888889   293.888889
2   20160603  68.4   NaN  55.6   NaN    NaN     20.222222   293.222222
1   20160602  65.8  80.8  55.0  25.8   25.8     18.777778   291.777778
28  20160629  65.8  73.2   NaN   NaN    NaN     18.777778   291.777778
29  20160630  65.7  72.7  59.2  13.5   13.5     18.722222   291.722222
24  20160625  65.7  75.4  57.9  17.5   17.5     18.722222   291.722222
0   20160601  65.5  73.6  54.7  18.9   18.9     18.611111   291.611111
27  20160628  65.4  73.0  55.8  17.2   17.2     18.555556   291.555556
14  20160615  63.4  70.3  49.3  21.0   21.0     17.444444   290.444444
20  20160621  62.6  71.4  50.4  21.0   21.0     17.000000   290.000000
21  20160622  61.7  70.2  55.4  14.8   14.8     16.500000   289.500000
23  20160624  61.1  68.9  56.7  12.2   12.2     16.166667   289.166667
22  20160623  60.9  67.1  54.9  12.2   12.2     16.055556   289.055556
26  20160627  60.7  70.0   NaN   NaN    NaN     15.944444   288.944444
16  20160617  60.4  70.7  55.9  14.8   14.8     15.777778   288.777778
13  20160614  59.7  67.8  47.8  20.0   20.0     15.388889   288.388889
19  20160620  59.3  69.1  52.2  16.9   16.9     15.166667   288.166667
12  20160613  58.3  68.2  47.3  20.9   20.9     14.611111   287.611111
15  20160616  57.8  67.5  55.6  11.9   11.9     14.333333   287.333333
3   20160604  57.5  70.9  47.3  23.6   23.6     14.166667   287.166667
17  20160618  57.3   NaN  54.0   NaN    NaN     14.055556   287.055556
6   20160607  56.9  65.1  45.9  19.2   19.2     13.833333   286.833333
18  20160619  56.3  59.2  54.1   5.1    5.1     13.500000   286.500000
11  20160612  55.4  64.2  46.0  18.2   18.2     13.000000   286.000000
7   20160608  54.2   NaN  47.5   NaN    NaN     12.333333   285.333333
10  20160611  54.0  62.1  41.7  20.4   20.4     12.222222   285.222222
5   20160606  52.2  59.7  42.8  16.9   16.9     11.222222   284.222222
4   20160605  51.4  58.3  43.2  15.1   15.1     10.777778   283.777778
9   20160610  49.5  55.9  43.0  12.9   12.9      9.722222   282.722222
8   20160609  49.4  54.1  45.7   8.4    8.4      9.666667   282.666667

# WRITING DATA TO A FILE

In [39]:
#define output file name
output_fp = "Kumpula_temps_June_2016.csv"

#save dataframe to csv
data.to_csv(output_fp, sep=",")
#the data produced has some columns with many decimal points

In [40]:
#define output file name
output_fp2 = "Kumpula_temps_above_15_June_2016.csv"

#save dataframe to csv
warm_temps.to_csv(output_fp2, sep=",", index=False, float_format="%.1f")
#removes index and floating point number written to one decimal point

# EXERCISE 5

# PROBLEM 1
Basic statistics of the data 

In [76]:
import pandas as pd
fp = ("6153237444115dat.csv")
#na_values = ["*", "**", "***", "****", "*****", "******"]

data = pd.read_csv(fp, na_values = ["*", "**", "***", "****", "*****", "******"])
data.head()

USAF   WBAN  YR--MODAHRMN    DIR   SPD   GUS  CLG  SKC   L   M  ...  \
0  28450  99999  201705010000  174.0  10.0  14.0  NaN  NaN NaN NaN  ...   
1  28450  99999  201705010020  180.0  10.0   NaN  4.0  NaN NaN NaN  ...   
2  28450  99999  201705010050  190.0  10.0   NaN  4.0  NaN NaN NaN  ...   
3  28450  99999  201705010100  188.0  12.0  16.0  NaN  NaN NaN NaN  ...   
4  28450  99999  201705010120  200.0  13.0   NaN  2.0  OBS NaN NaN  ...   

      SLP    ALT    STP  MAX  MIN  PCP01  PCP06  PCP24  PCPXX    SD  
0  1009.2    NaN  984.1  NaN  NaN    NaN    NaN    NaN    NaN  35.0  
1     NaN  29.74    NaN  NaN  NaN    NaN    NaN    NaN    NaN   NaN  
2     NaN  29.74    NaN  NaN  NaN    NaN    NaN    NaN    NaN   NaN  
3  1009.1    NaN  984.0  NaN  NaN    NaN    NaN    NaN    NaN  35.0  
4     NaN  29.74    NaN  NaN  NaN    NaN    NaN    NaN    NaN   NaN  

[5 rows x 33 columns]

In [42]:
row_count = len(data)
column_names = data.columns.values
column_datatypes = data.dtypes

In [43]:
# Print the number of rows in the dataframe:
print(f"There are {row_count} rows")

There are 11694 rows


In [44]:
# Print the column names:
print(f"The columns are: \n{column_names}")

The columns are: 
['USAF' 'WBAN' 'YR--MODAHRMN' 'DIR' 'SPD' 'GUS' 'CLG' 'SKC' 'L' 'M' 'H'
 'VSB' 'MW' 'MW.1' 'MW.2' 'MW.3' 'AW' 'AW.1' 'AW.2' 'AW.3' 'W' 'TEMP'
 'DEWP' 'SLP' 'ALT' 'STP' 'MAX' 'MIN' 'PCP01' 'PCP06' 'PCP24' 'PCPXX' 'SD']


In [45]:
# Print the column datatypes:
print(f"The column types are: \n{column_datatypes}")

The column types are: 
USAF              int64
WBAN              int64
YR--MODAHRMN      int64
DIR             float64
SPD             float64
GUS             float64
CLG             float64
SKC              object
L               float64
M               float64
H               float64
VSB             float64
MW              float64
MW.1            float64
MW.2            float64
MW.3            float64
AW              float64
AW.1            float64
AW.2            float64
AW.3            float64
W               float64
TEMP            float64
DEWP            float64
SLP             float64
ALT             float64
STP             float64
MAX             float64
MIN             float64
PCP01           float64
PCP06           float64
PCP24           float64
PCPXX           float64
SD              float64
dtype: object


In [46]:
#calculate mean temp in fahr
temp_mean = data["TEMP"].mean()

# Calculate standard deviaton of the maximum temperature:
temp_max_std = data["MAX"].std()

#Calculate number of unique stations:
station_count = len(data["USAF"].unique())

In [47]:
# Check mean temperature value
print(f"The mean temperature in Fahrenheit is {round(temp_mean,1)}")

The mean temperature in Fahrenheit is 52.2


In [48]:
# Check standard deviation value
print(f"The standard deviation of maximum temperature is {round(temp_max_std, 1)}")

The standard deviation of maximum temperature is 10.3


In [49]:
# Check number of stations value
print(f"The number of unique stations is {station_count}")

The number of unique stations is 2


# PROBLEM 2
Data manipulation and selection

In [50]:
#select number of columns
selected = data[["USAF", "YR--MODAHRMN", "TEMP", "MAX", "MIN"]]
#emove all rows from selected that have NoData in the column TEMP
selected = selected.dropna(subset=["TEMP"])

In [51]:
selected.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN
0  28450  201705010000  31.0  NaN  NaN
1  28450  201705010020  30.0  NaN  NaN
2  28450  201705010050  30.0  NaN  NaN
3  28450  201705010100  31.0  NaN  NaN
4  28450  201705010120  30.0  NaN  NaN

In [52]:
#Check how many rows you have after removing the no-data values:
len(selected)

11691

In [53]:
# creates new column Celsius and adds fahrenheits converted to celsius in it
selected["Celsius"] = (selected["TEMP"] - 32) / 1.8
# round values in column Celsius to integer
selected["Celsius"] = selected["Celsius"].round(0).astype(int)

In [54]:
selected.head()

USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1

In [55]:
# Check data types
print(selected.dtypes)

USAF              int64
YR--MODAHRMN      int64
TEMP            float64
MAX             float64
MIN             float64
Celsius           int32
dtype: object


In [56]:
#Select all rows from the selected DataFrame with the USAF code 29980 into a variable called kumpula
#Select all rows from the selected DataFrame with the USAF code 28450 into a variable called rovaniemi

kumpula = selected.loc[selected["USAF"] == 29980]
rovaniemi = selected.loc[selected["USAF"] == 28450]

#reset index
kumpula = kumpula.reset_index(drop=True)
rovaniemi = rovaniemi.reset_index(drop=True)


print(f"Kumpula: \n{kumpula.head()}\n")
print(f"Rovaniemi: \n{rovaniemi.head()}\n")

Kumpula: 
    USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  29980  201705010000  37.0  NaN  NaN        3
1  29980  201705010100  37.0  NaN  NaN        3
2  29980  201705010200  37.0  NaN  NaN        3
3  29980  201705010300  37.0  NaN  NaN        3
4  29980  201705010400  39.0  NaN  NaN        4

Rovaniemi: 
    USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1



In [57]:
#save the kumpula DataFrame into the file Kumpula_temps_May_Aug_2017.csv
output_kum = ("Kumpula_temps_May_Aug_2017.csv")
kumpula.to_csv(output_kum, sep=",", float_format="%.2f")

#Save the rovaniemi DataFrame into the file Rovaniemi_temps_May_Aug_2017.csv
output_rov = ("Rovaniemi_temps_May_August_2017.csv")
rovaniemi.to_csv(output_rov, sep=",", float_format="%.2f")

# PROBLEM 3
Data analysis

In [58]:
#Read in the csv files generated in Problem 2 to the variables kumpula and rovaniemi
import pandas as pd
kumpula = pd.read_csv("Kumpula_temps_May_Aug_2017.csv") 
romaniemi = pd.read_csv("Rovaniemi_temps_May_August_2017.csv")

print(kumpula.head())
print()
print(rovaniemi.head())

   Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0           0  29980  201705010000  37.0  NaN  NaN        3
1           1  29980  201705010100  37.0  NaN  NaN        3
2           2  29980  201705010200  37.0  NaN  NaN        3
3           3  29980  201705010300  37.0  NaN  NaN        3
4           4  29980  201705010400  39.0  NaN  NaN        4

    USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1


In [59]:
#What was the median Celsius temperature during the observed period
kumpula_median = kumpula["Celsius"].median()
rovaniemi_median = rovaniemi["Celsius"].median()

print(f"Kumpula median is: {kumpula_median}")
print(f"Rovaniemi median is: {rovaniemi_median}")

Kumpula median is: 14.0
Rovaniemi median is: 11.0


In [60]:
#From the kumpula and rovaniemi DataFrames, select the rows where values of the YR--MODAHRMN column are from May 2017
#Assign these selected rows into the variables kumpula_may and rovaniemi_may (you can check the hints for help!)
#Repeat the procedure for the month of June and assign those values into variables to kumpula_june and rovaniemi_june

kumpula_may = kumpula.loc[(kumpula["YR--MODAHRMN"] >= 201705010000) & (kumpula["YR--MODAHRMN"] <= 201705312359)]
rovaniemi_may = rovaniemi.loc[(rovaniemi["YR--MODAHRMN"] >= 201705010000) & (rovaniemi["YR--MODAHRMN"] < 201706010000)]

kumpula_june = kumpula.loc[(kumpula["YR--MODAHRMN"] >= 201706010000) & (kumpula["YR--MODAHRMN"] < 201707010000)]
rovaniemi_june = rovaniemi.loc[(rovaniemi["YR--MODAHRMN"] >= 201706010000) & (rovaniemi["YR--MODAHRMN"] <= 201707010000)]

In [61]:
print(f"First values in May, Kumpula:\n{kumpula_may.head()}\n")
print(f"Last values in May, Kumpula:\n{kumpula_may.tail()}")

First values in May, Kumpula:
   Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0           0  29980  201705010000  37.0  NaN  NaN        3
1           1  29980  201705010100  37.0  NaN  NaN        3
2           2  29980  201705010200  37.0  NaN  NaN        3
3           3  29980  201705010300  37.0  NaN  NaN        3
4           4  29980  201705010400  39.0  NaN  NaN        4

Last values in May, Kumpula:
     Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
736         736  29980  201705311900  51.0  NaN  NaN       11
737         737  29980  201705312000  50.0  NaN  NaN       10
738         738  29980  201705312100  47.0  NaN  NaN        8
739         739  29980  201705312200  44.0  NaN  NaN        7
740         740  29980  201705312300  43.0  NaN  NaN        6


In [62]:
print(f"First values in June, Kumpula:\n{kumpula_june.head()}\n")
print(f"Last values in June, Kumpula:\n{kumpula_june.tail()}")

First values in June, Kumpula:
     Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
741         741  29980  201706010000  42.0  NaN  NaN        6
742         742  29980  201706010100  40.0  NaN  NaN        4
743         743  29980  201706010200  40.0  NaN  NaN        4
744         744  29980  201706010300  41.0  NaN  NaN        5
745         745  29980  201706010400  44.0  NaN  NaN        7

Last values in June, Kumpula:
      Unnamed: 0   USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
1450        1450  29980  201706301900  65.0  NaN  NaN       18
1451        1451  29980  201706302000  61.0  NaN  NaN       16
1452        1452  29980  201706302100  63.0  NaN  NaN       17
1453        1453  29980  201706302200  62.0  NaN  NaN       17
1454        1454  29980  201706302300  61.0  NaN  NaN       16


In [63]:
print(f"First values in May, Rovaniemi:\n{rovaniemi_may.head()}\n")
print(f"Last values in May, Rovaniemi:\n{rovaniemi_may.tail()}")

First values in May, Rovaniemi:
    USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
0  28450  201705010000  31.0  NaN  NaN       -1
1  28450  201705010020  30.0  NaN  NaN       -1
2  28450  201705010050  30.0  NaN  NaN       -1
3  28450  201705010100  31.0  NaN  NaN       -1
4  28450  201705010120  30.0  NaN  NaN       -1

Last values in May, Rovaniemi:
       USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
2215  28450  201705312220  32.0  NaN  NaN        0
2216  28450  201705312250  32.0  NaN  NaN        0
2217  28450  201705312300  33.0  NaN  NaN        1
2218  28450  201705312320  32.0  NaN  NaN        0
2219  28450  201705312350  30.0  NaN  NaN       -1


In [64]:
print(f"First values in June, Rovaniemi:\n{rovaniemi_june.head()}\n")
print(f"Last values in June, Rovaniemi:\n{rovaniemi_june.tail()}")

First values in June, Rovaniemi:
       USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
2220  28450  201706010000  32.0  NaN  NaN        0
2221  28450  201706010020  30.0  NaN  NaN       -1
2222  28450  201706010050  30.0  NaN  NaN       -1
2223  28450  201706010100  31.0  NaN  NaN       -1
2224  28450  201706010120  30.0  NaN  NaN       -1

Last values in June, Rovaniemi:
       USAF  YR--MODAHRMN  TEMP  MAX  MIN  Celsius
4343  28450  201706302250  57.0  NaN  NaN       14
4344  28450  201706302300  56.0  NaN  NaN       13
4345  28450  201706302320  57.0  NaN  NaN       14
4346  28450  201706302350  57.0  NaN  NaN       14
4347  28450  201707010000  56.0  NaN  NaN       13


In [65]:
print("Kumpula May- Mean: " + str(round(kumpula_may["Celsius"].mean())) + " Min: " + str(kumpula_may["Celsius"].min()) + " Max: " + str(kumpula_may["Celsius"].max()))
print("Kumpula June- Mean: " + str(round(kumpula_june["Celsius"].mean())) + " Min: " + str(kumpula_june["Celsius"].min()) + " Max: " + str(kumpula_june["Celsius"].max()))
print("Rovaniemi May- Mean: " + str(round(rovaniemi_may["Celsius"].mean())) + " Min: " + str(rovaniemi_may["Celsius"].min()) + " Max: " + str(rovaniemi_may["Celsius"].max()))
print("Rovaniemi June- Mean: " + str(round(rovaniemi_june["Celsius"].mean())) + " Min: " + str(rovaniemi_june["Celsius"].min()) + " Max: " + str(rovaniemi_june["Celsius"].max()))

Kumpula May- Mean: 10 Min: -2 Max: 23
Kumpula June- Mean: 14 Min: 3 Max: 24
Rovaniemi May- Mean: 3 Min: -7 Max: 15
Rovaniemi June- Mean: 11 Min: -1 Max: 23


# Problem 4
Parsing daily temperatures

## Data Frame Columns Part by Part

In [101]:
#import pandas
import pandas as pd

#read data
fp = ('6153237444115dat.csv')

data = pd.read_csv(fp,
                   na_values = ['*', '**', '***', '****', '*****', '******'],
                   usecols = ['USAF', 'YR--MODAHRMN', 'TEMP', 'MAX', 'MIN'],
                   parse_dates = ['YR--MODAHRMN'], #converts date to date time
                   #index_col = 'YR--MODAHRMN', #set date as index
)

data.head()

USAF        YR--MODAHRMN  TEMP  MAX  MIN
0  28450 2017-05-01 00:00:00  31.0  NaN  NaN
1  28450 2017-05-01 00:20:00  30.0  NaN  NaN
2  28450 2017-05-01 00:50:00  30.0  NaN  NaN
3  28450 2017-05-01 01:00:00  31.0  NaN  NaN
4  28450 2017-05-01 01:20:00  30.0  NaN  NaN

In [106]:
#rename columns
data.rename(columns={'YR--MODAHRMN':'DATETIME', 'USAF':'STATION'}, inplace=True)

# Convert to datetime and keep only year and month
data['DATE'] = data['DATETIME'].dt.date

#drop na values
data= data.dropna(subset=['TEMP'])

# creates new column Celsius and adds fahrenheits converted to celsius in it
data['TEMP_C'] = (data["TEMP"] - 32) / 1.8
# round values in column Celsius to integer
data['TEMP_C'] = data['TEMP_C'].round(2).astype(int)

#set date column as index
#data.set_index('YR_MN_DT', inplace=True)
              
data.head()

STATION            DATETIME  TEMP  MAX  MIN        DATE  TEMP_C
0    28450 2017-05-01 00:00:00  31.0  NaN  NaN  2017-05-01       0
1    28450 2017-05-01 00:20:00  30.0  NaN  NaN  2017-05-01      -1
2    28450 2017-05-01 00:50:00  30.0  NaN  NaN  2017-05-01      -1
3    28450 2017-05-01 01:00:00  31.0  NaN  NaN  2017-05-01       0
4    28450 2017-05-01 01:20:00  30.0  NaN  NaN  2017-05-01      -1

In [107]:
#group by day
day_grouped = data.groupby('DATE')

In [108]:
len(data)

11691

In [109]:
len(day_grouped)

123

In [110]:
day_grouped.groups.keys()

dict_keys([datetime.date(2017, 5, 1), datetime.date(2017, 5, 2), datetime.date(2017, 5, 3), datetime.date(2017, 5, 4), datetime.date(2017, 5, 5), datetime.date(2017, 5, 6), datetime.date(2017, 5, 7), datetime.date(2017, 5, 8), datetime.date(2017, 5, 9), datetime.date(2017, 5, 10), datetime.date(2017, 5, 11), datetime.date(2017, 5, 12), datetime.date(2017, 5, 13), datetime.date(2017, 5, 14), datetime.date(2017, 5, 15), datetime.date(2017, 5, 16), datetime.date(2017, 5, 17), datetime.date(2017, 5, 18), datetime.date(2017, 5, 19), datetime.date(2017, 5, 20), datetime.date(2017, 5, 21), datetime.date(2017, 5, 22), datetime.date(2017, 5, 23), datetime.date(2017, 5, 24), datetime.date(2017, 5, 25), datetime.date(2017, 5, 26), datetime.date(2017, 5, 27), datetime.date(2017, 5, 28), datetime.date(2017, 5, 29), datetime.date(2017, 5, 30), datetime.date(2017, 5, 31), datetime.date(2017, 6, 1), datetime.date(2017, 6, 2), datetime.date(2017, 6, 3), datetime.date(2017, 6, 4), datetime.date(2017, 6,

In [112]:
#iterate over groups
for key, group in day_grouped:
    #print key and group
    print(f"Key:\n {key}")
    print(f"\n First rows of data in this group:\n {group.head()}")
    
    #stop iteration
    break

Key:
 2017-05-01

 First rows of data in this group:
    STATION            DATETIME  TEMP  MAX  MIN        DATE  TEMP_C
0    28450 2017-05-01 00:00:00  31.0  NaN  NaN  2017-05-01       0
1    28450 2017-05-01 00:20:00  30.0  NaN  NaN  2017-05-01      -1
2    28450 2017-05-01 00:50:00  30.0  NaN  NaN  2017-05-01      -1
3    28450 2017-05-01 01:00:00  31.0  NaN  NaN  2017-05-01       0
4    28450 2017-05-01 01:20:00  30.0  NaN  NaN  2017-05-01      -1


In [116]:
# create empty dataframe for aggregated values
daily_data = pd.DataFrame()

#columns to aggregate
mean_cols = ["TEMP", "MIN", "MAX", "TEMP_C"]

#iterate over the groups
for key, group in day_grouped:
    #calc mean
    mean_values = group[mean_cols].mean()
    
    # Add the ´key´ (i.e. the date+time information) into the aggregated values
    mean_values["DATE"] = key
    
    # Convert the mean_values series to a DataFrame and make it have a row orientation
    row = mean_values.to_frame().transpose()
    
    # Concatenate the aggregated values into the monthly_data DataFrame
    daily_data = pd.concat([daily_data, row], ignore_index=True)

In [118]:
daily_data

TEMP    MIN    MAX     TEMP_C        DATE
0    38.354167  34.75   44.0   3.260417  2017-05-01
1         41.0   38.5   49.5   4.583333  2017-05-02
2    39.021053  38.75   46.5   3.463158  2017-05-03
3    40.979167   36.5   47.0   4.708333  2017-05-04
4     45.96875   40.0  52.75   7.333333  2017-05-05
..         ...    ...    ...        ...         ...
118  47.221053  45.25   51.5   7.968421  2017-08-27
119  49.705263  46.75  54.25   9.421053  2017-08-28
120  53.072917  49.75  56.75  11.291667  2017-08-29
121   54.78125  54.75  60.25    12.3125  2017-08-30
122  55.854167   54.5   61.0      12.75  2017-08-31

[123 rows x 5 columns]

## Full Solution

In [1]:
#import pandas
import pandas as pd

#read data
fp = ('6153237444115dat.csv')

data = pd.read_csv(fp,
                   na_values = ['*', '**', '***', '****', '*****', '******'],
                   usecols = ['USAF', 'YR--MODAHRMN', 'TEMP', 'MAX', 'MIN'],
                   parse_dates = ['YR--MODAHRMN'], #converts date to date time
)

#rename columns
data.rename(columns={'YR--MODAHRMN':'DATETIME', 'USAF':'STATION'}, inplace=True)

# Create column from datetime for yr-mn-dt
data['DATE'] = data['DATETIME'].dt.date

#drop na values
data= data.dropna(subset=['TEMP'])

# creates new column Celsius and adds fahrenheits converted to celsius in it
data['TEMP_C'] = (data["TEMP"] - 32) / 1.8
# round values in column Celsius to integer
data['TEMP_C'] = data['TEMP_C'].round(2).astype(int)

#create groups by day
day_grouped = data.groupby('DATE')

# create empty dataframe for aggregated values
daily_data = pd.DataFrame()

#columns to aggregate
mean_cols = ["TEMP", "MIN", "MAX", "TEMP_C"]

#iterate over the groups
for key, group in day_grouped:
    #calc mean
    mean_values = group[mean_cols].mean()
    
    # Add the ´key´ (i.e. the date+time information) into the aggregated values
    mean_values["DATE"] = key
    
    # Convert the mean_values series to a DataFrame and make it have a row orientation
    row = mean_values.to_frame().transpose()
    
    # Concatenate the aggregated values into the monthly_data DataFrame
    daily_data = pd.concat([daily_data, row], ignore_index=True)
    
daily_data

TEMP    MIN    MAX     TEMP_C        DATE
0    38.354167  34.75   44.0   3.260417  2017-05-01
1         41.0   38.5   49.5   4.583333  2017-05-02
2    39.021053  38.75   46.5   3.463158  2017-05-03
3    40.979167   36.5   47.0   4.708333  2017-05-04
4     45.96875   40.0  52.75   7.333333  2017-05-05
..         ...    ...    ...        ...         ...
118  47.221053  45.25   51.5   7.968421  2017-08-27
119  49.705263  46.75  54.25   9.421053  2017-08-28
120  53.072917  49.75  56.75  11.291667  2017-08-29
121   54.78125  54.75  60.25    12.3125  2017-08-30
122  55.854167   54.5   61.0      12.75  2017-08-31

[123 rows x 5 columns]

## Solution

In [93]:
#import pandas
import pandas as pd

#read data
fp = ('6153237444115dat.csv')

data = pd.read_csv(fp,
                   na_values = ['*', '**', '***', '****', '*****', '******'],
                   usecols = ['USAF', 'YR--MODAHRMN', 'TEMP', 'MAX', 'MIN'],
                   parse_dates = ['YR--MODAHRMN'], #converts date to date time
)

#rename columns
data.rename(columns={'YR--MODAHRMN':'DATETIME', 'USAF':'STATION'}, inplace=True)

# Create column from datetime for yr-mn-dt
data['DATE'] = data['DATETIME'].dt.date

#drop na values
data= data.dropna(subset=['TEMP'])

# creates new column Celsius and adds fahrenheits converted to celsius in it
data['TEMP_C'] = (data["TEMP"] - 32) / 1.8
# round values in column Celsius to integer
data['TEMP_C'] = data['TEMP_C'].round(2).astype(int)

#create groups by day
day_grouped = data.groupby('DATE')

#Get mean, max and min values
mean_val = day_grouped['TEMP_C'].mean()
max_val = day_grouped['TEMP_C'].max()
min_val = day_grouped['TEMP_C'].min()

#add them to a new dataframe and rename the columns
daily_data = pd.concat([mean_val, max_val, min_val], axis=1)
daily_data.columns = ['Daily Mean', 'Daily Max', 'Daily Min']
daily_data

Daily Mean  Daily Max  Daily Min
DATE                                        
2017-05-01    3.260417         12         -1
2017-05-02    4.583333         15          0
2017-05-03    3.463158         13         -1
2017-05-04    4.708333         10         -1
2017-05-05    7.333333         16          1
...                ...        ...        ...
2017-08-27    7.968421         14          5
2017-08-28    9.421053         16          2
2017-08-29   11.291667         17          7
2017-08-30   12.312500         18          8
2017-08-31   12.750000         18          7

[123 rows x 3 columns]

## For Kumpula and Rovaniem Data

In [101]:
#import pandas
import pandas as pd

#Read in the data
kumpula = pd.read_csv('Kumpula_temps_May_Aug_2017.csv',
                      parse_dates = ['YR--MODAHRMN']
                     )

rovaniemi = pd.read_csv('Rovaniemi_temps_May_August_2017.csv',
                        parse_dates = ['YR--MODAHRMN']
                       )
#delete columns
del(kumpula['Unnamed: 0'])
del(rovaniemi['Unnamed: 0'])

#add date column
kumpula['DATE'] = kumpula['YR--MODAHRMN'].dt.date
rovaniemi['DATE'] = rovaniemi['YR--MODAHRMN'].dt.date

#group by date
kumpula_date_group = kumpula.groupby('DATE')
rovaniemi_date_group = rovaniemi.groupby('DATE')

#calculate min, max and mean
kumpula_mean = kumpula_date_group['Celsius'].mean()
kumpula_max = kumpula_date_group['Celsius'].max()
kumpula_min = kumpula_date_group['Celsius'].min()

rovaniemi_mean = rovaniemi_date_group['Celsius'].mean()
rovaniemi_max = rovaniemi_date_group['Celsius'].max()
rovaniemi_min = rovaniemi_date_group['Celsius'].min()

#add to new data frame
kumpula_data = pd.concat([kumpula_mean, kumpula_max, kumpula_min], axis=1)
kumpula_data.columns = ['Kumpula Mean', 'Kumpula Max', 'Kumpula Min']

rovaniemi_data = pd.concat([rovaniemi_mean, rovaniemi_max, rovaniemi_min], axis=1)
rovaniemi_data.columns = ['Rovaniemi Mean', 'Rovaniemi Max', 'Rovaniemi Min']

In [102]:
kumpula_data

Kumpula Mean  Kumpula Max  Kumpula Min
DATE                                              
2017-05-01      7.625000           12            3
2017-05-02      9.750000           16            2
2017-05-03      9.208333           13            4
2017-05-04      6.666667           11            3
2017-05-05     10.250000           17            2
...                  ...          ...          ...
2017-08-27     10.625000           14            6
2017-08-28     11.826087           16            9
2017-08-29     14.500000           17            8
2017-08-30     16.833333           19           15
2017-08-31     17.250000           19           16

[123 rows x 3 columns]

In [103]:
rovaniemi_data

Rovaniemi Mean  Rovaniemi Max  Rovaniemi Min
DATE                                                    
2017-05-01        2.180556              7             -1
2017-05-02        3.402778              7              1
2017-05-03        2.112676              4             -1
2017-05-04        4.388889              9             -1
2017-05-05        6.916667             12              1
...                    ...            ...            ...
2017-08-27        7.690141             10              5
2017-08-28        9.138889             13              3
2017-08-29       10.722222             12              8
2017-08-30       11.291667             14              9
2017-08-31       12.000000             17              8

[123 rows x 3 columns]

In [104]:
kumpula

USAF        YR--MODAHRMN  TEMP  MAX  MIN  Celsius        DATE
0     29980 2017-05-01 00:00:00  37.0  NaN  NaN        3  2017-05-01
1     29980 2017-05-01 01:00:00  37.0  NaN  NaN        3  2017-05-01
2     29980 2017-05-01 02:00:00  37.0  NaN  NaN        3  2017-05-01
3     29980 2017-05-01 03:00:00  37.0  NaN  NaN        3  2017-05-01
4     29980 2017-05-01 04:00:00  39.0  NaN  NaN        4  2017-05-01
...     ...                 ...   ...  ...  ...      ...         ...
2919  29980 2017-08-31 19:00:00  64.0  NaN  NaN       18  2017-08-31
2920  29980 2017-08-31 20:00:00  64.0  NaN  NaN       18  2017-08-31
2921  29980 2017-08-31 21:00:00  64.0  NaN  NaN       18  2017-08-31
2922  29980 2017-08-31 22:00:00  64.0  NaN  NaN       18  2017-08-31
2923  29980 2017-08-31 23:00:00  64.0  NaN  NaN       18  2017-08-31

[2924 rows x 7 columns]

In [105]:
#alternative method
kumpula_data2 = kumpula.groupby('DATE').Celsius.agg(['mean','max','min'])
rovaniemi_data2 = rovaniemi.groupby('DATE').Celsius.agg(['mean','max','min'])

In [106]:
kumpula_data2

mean  max  min
DATE                           
2017-05-01   7.625000   12    3
2017-05-02   9.750000   16    2
2017-05-03   9.208333   13    4
2017-05-04   6.666667   11    3
2017-05-05  10.250000   17    2
...               ...  ...  ...
2017-08-27  10.625000   14    6
2017-08-28  11.826087   16    9
2017-08-29  14.500000   17    8
2017-08-30  16.833333   19   15
2017-08-31  17.250000   19   16

[123 rows x 3 columns]